File: nadalgo/no_tiebreak/montecarlo.py
import random

DEBUG = 1

def simu(nb_simu, p=0.55, nb_sets=2):
    wins = 0
    for _ in range(nb_simu):
        scores = [0, 0]
        s1, s2 = scores
        while s1 < nb_sets and s2 < nb_sets:
            scores[set(p)] += 1
            s1, s2 = scores
        if s1 > s2:
            wins += 1

    return wins / nb_simu

def template(p, upto, func, repeat=10000):
    scores = [0, 0]
    for _ in range(repeat):
        scores[func(p)] += 1
        s1, s2 = scores
        if s1 >= upto and s1 - s2 >= 2:
            return 0
        elif s2 >= upto and s2 - s1 >= 2:
            return 1

def jeu(p):
    return template(p, 4, lambda p: 0 if random.random() <= p else 1)

def set(p):
    return template(p, 6, jeu)

if DEBUG:
    print(simu(100000, p=0.55, nb_sets=2)) # 0.95771
    print(simu(50000, p=0.55, nb_sets=2)) # 0.95802
    print(simu(100000, p=0.6, nb_sets=2)) # 0.99969

File: puzzles/nadalgo/no_tiebreak/montecarlo.cpp
#include <iostream>
#include <random>

using namespace std;


std::random_device rd;
std::mt19937 gen(rd());
std::uniform_real_distribution<> dis(0.0, 1.0);


int _template(double p, int upto, int (*func)(double), int repeat = 10000) {
    int scores[2] = {0, 0};
    for (int i = 0; i < repeat; i++) {
        scores[func(p)] += 1;
        int s1 = scores[0];
        int s2 = scores[1];
        if (s1 >= upto && s1 - s2 >= 2) {
            return 0;
        } else if (s2 >= upto && s2 - s1 >= 2) {
            return 1;
        }
    }
}

int jeu(double p) {
    return _template(p, 4, [](double p) { return (dis(gen) <= p) ? 0 : 1; });
}

int _set(double p) {
    return _template(p, 6, jeu);
}

double simu(int nb_simu, double p = 0.55, int nb_sets = 2) {
    int wins = 0;
    for (int i = 0; i < nb_simu; i++) {
        int scores[2] = {0, 0};
        int s1 = scores[0];
        int s2 = scores[1];
        while (s1 < nb_sets && s2 < nb_sets) {
            scores[_set(p)] += 1;
            s1 = scores[0];
            s2 = scores[1];
        }
        if (s1 > s2) {
            wins += 1;
        }
    }

    return double(wins) / nb_simu;
}


int main() {
    float p;
    int n_sets;
    bool input = false;
    if (input) {
        cin >> p >> n_sets; cin.ignore();
        cout << simu(5000, p, n_sets);
    } else {
        cout << simu(100000, 0.55, 2) << endl;
        cout << simu(50000, 0.55, 2) << endl;
        cout << simu(100000, 0.6, 2) << endl;
    }
    
    return 0;
}
